core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -88,6 +93,10 @@
   private boolean fullStreamDecompression;
   private DecompressorRegistry decompressorRegistry = DecompressorRegistry.getDefaultInstance();
   private CompressorRegistry compressorRegistry = CompressorRegistry.getDefaultInstance();
+  private volatile ScheduledFuture<?> deadlineCancellationNotifyApplicationFuture;
+  private volatile ScheduledFuture<?> deadlineCancellationSendToServerFuture;
+  private boolean observerClosed = false;
+  private volatile boolean streamCancelledByDeadline = false;
```

Comment:
`stream.cancel()` is safe to be called more than once. This boolean is unnecessary.